pandas/core/arrays/string_.py

```diff
@@ -410,6 +411,12 @@ def __setitem__(self, key, value):
 
         super().__setitem__(key, value)
 
+    def _putmask(self, mask: npt.NDArray[np.bool_], value) -> None:
+        # the super() method NDArrayBackedExtensionArray._putmask uses
+        # np.putmask which doesn't properly handle None/pd.NA, so using the
+        # base class implementation that uses __setitem__
+        ExtensionArray._putmask(self, mask, value)
```

Comment:
Is it worth dispatching this to `ExtensionArray._putmask` if `None/pd.NA` is set for perf reasons?

Reply:
I don't fully understand your question, I think? (what do you mean with "for perf reasons"?)

The base class is basically just doing `self[mask] = value`, and `StringArray.__setitem__` (just above) handles validation of the `value` (something that np.putmask doesn't do)

Reply:
e.g.

```
if value is None/pd.NA:
    ExtensionArray._putmask(self, mask, value)
else:
    # More performant than ExtensionArray._putmask(?) but returns invalid results for None/pd.NA
    super()._putmask(mask, value)
```

Reply:
Ah, I see. It's not only None/pd.NA (although this was the original bug report), but it turned out to also not work for any non-string (see the tests I added in https://github.com/pandas-dev/pandas/pull/47763/files#diff-7acefeb50e80f3dee3fceec56c42dc4a37596d9a119930d5a3af2b44c1b1dcbbR603-R610). 

So we could do a check like you propose, but then it would be something like:

```
if isinstance(value, str):
    # using np.putmask
    super()._putmask(mask, value)
else:
    # fallback to base implementation using setitem
    ExtensionArray._putmask(self, mask, value)
```

But I am not sure it is worth the extra complexity. Will do a quick check to see the difference in performance.

Reply:
The putmask version doesn't seem to be faster (for this object dtype case):

```
In [19]: arr = pd.array(["a", "b", "c", None]*10000, dtype="string")

# convert to bool numpy array to have the timing focus on the actual setting
In [20]: mask = (arr == "a").fillna(False).to_numpy(bool)

In [21]: %timeit arr[mask] = "a"
86.7 µs ± 20.9 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [22]: %timeit arr._putmask(mask, "a")
99.8 µs ± 4.42 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

So in that case we can keep a single version dispatching to the base class one I think